pyiceberg/table/__init__.py

```diff
@@ -1156,7 +1166,9 @@ def overwrite(
         if len(self.spec().fields) > 0:
             raise ValueError("Cannot write to partitioned tables")
 
-        _check_schema(self.schema(), other_schema=df.schema)
+        _check_schema_compatible(self.schema(), other_schema=df.schema)
+        # the two schemas are compatible so safe to cast
+        df = df.cast(self.schema().as_arrow())
```

Comment:
Should `_check_schema_compatible` return a bool to indicate if the cast is needed? I'm not sure how costly the cast is. If we go from `string` to `large_string` then we might rewrite the Arrow buffers.

Reply:
yea, I like that idea. 
`_check_schema_compatible` returns a boolean `should_cast`.
* If schema is exactly the same, return `False` and skip cast
* If schema is "compatible", return `True` and cast
* If schema is not "compatible", throws an error

Reply:
It was too complicated when `_check_schema_compatible` returned a boolean and threw an error. 
I ended up doing an extra comparison as Arrow schemas outside and cast only if necessary